### PR TITLE
Logistic regression

### PR DESCRIPTION
**Pull request by MrinankSharma**

### CODE REVIEW

File: src/model/logistic_regression_models.py [hunk 1]
```diff
@@ -176,61 +268,36 @@ def predict(self, x, integration_limit=500, parameters=None, hyperparameters=Non
...
+        :param hyperparameters:
+        :return: optimum params
+        """
+        super().fit(data, t_i, parameters, hyperparameters)
 
-    def update_hyperparameters(self, updated_dict):
-        for key, value in updated_dict.items():
-            self.__dict__[key] = value
+        # convert data into a tensor
+        X = torch.tensor(data["X"], dtype=torch.float64)
```
Comment (MJHutchinson) [inline src/model/logistic_regression_models.py]: Default terminology has been for the x key to be lowercase in dictionaries

File: src/privacy_accounting/analysis/privacy_ledger.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def record_sum_query(self, l2_clipping_bound, noise_stddev):
         self._query_count = self._query_count + 1
         self._query_buffer.append(torch.Tensor([self._sample_count, l2_clipping_bound, noise_stddev]))
 
-    def finalise_sample(self):
```
Comment (MJHutchinson) [inline src/privacy_accounting/analysis/privacy_ledger.py]: This should be sample - a minibatch is a single sample you draw from the data set to apply a query to.

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 3]
```diff
@@ -53,13 +53,17 @@ def process_microbatch(losses, sample_state):
...
 
         self.apply_grads(param_groups, grads=final_grads)
 
-        self.optimiser.step()
-        self.optimiser.zero_grad()
+        self.optimizer.step()
+        self.optimizer.zero_grad()
+
+        # compute the new loss
+        current_loss = torch.sum(self.loss_per_example(self.model(x), y))
```
Comment (MJHutchinson) [inline src/privacy_accounting/optimizer/dp_optimizer.py]: does this quantity not exist already? Just sum it from the original loss call rather than calling the model twice

File: src/privacy_accounting/optimizer/standard_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        # note that this is a function handle
+        self.loss_per_example = loss_per_example
+
+    def fit_batch(self, X: torch.Tensor, y: torch.Tensor):
+        loss = self.loss_per_example(self.model(X), y)
+        total_loss = torch.sum(loss)
+
+        self.optimizer.zero_grad()
+        # to investigate whether this actually needs to be true
+        total_loss.backward(retain_graph=True)
```
Comment (MJHutchinson) [inline src/privacy_accounting/optimizer/standard_optimizer.py]: This should not need to be true. Needed in the micro batching case but should definitely not be here.

File: src/utils/torch_tensor_buffer.py [hunk 5]
```diff
@@ -45,7 +47,7 @@ def _double_capacity():
         assert value.shape == self._entry_shape
 
         self._buffer[self._current_size, :] = value
-        self._current_size = self._current_size + 1
+        self._current_size = self.current_size + 1
```
Comment (MJHutchinson) [inline src/utils/torch_tensor_buffer.py]: Supposed to be incrementing a variable - instead this accesses a variable that doesnt exist?

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 3 above)
Comment (MrinankSharma) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: the model parameters have changed - granted though this is slightly inefficient.

File: src/utils/torch_tensor_buffer.py (hunk 5 above)
Comment (MrinankSharma) [reply src/utils/torch_tensor_buffer.py]: typo - not sure how this happened tbh
